### PR TITLE
FEATURE: RAIL-423 Examples Testcafe infrastructure

### PR DESCRIPTION
**Pull request by tomasreichmann**

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/556/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/462/ : SUCCESS in 2m 50s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/558/ : SUCCESS in 10s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/465/ : SUCCESS in 58s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/559/ : SUCCESS in 20s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/466/ : SUCCESS in 46s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/560/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/467/ : SUCCESS in 51s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/561/ : SUCCESS in 7s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/468/ : SUCCESS in 2m 00s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/565/ : SUCCESS in 12s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/472/ : SUCCESS in 42s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/571/ : SUCCESS in 10s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/474/ : SUCCESS in 42s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/577/ : SUCCESS in 12s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/478/ : SUCCESS in 2m 28s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/579/ : SUCCESS in 7s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/481/ : SUCCESS in 39s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/583/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/485/ : SUCCESS in 2m 07s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/585/ : SUCCESS in 7s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/488/ : SUCCESS in 1m 58s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/586/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/489/ : SUCCESS in 3m 25s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/587/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/490/ : SUCCESS in 46s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/593/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/495/ : SUCCESS in 50s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/594/ : SUCCESS in 11s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/496/ : SUCCESS in 51s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/595/ : SUCCESS in 30s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/497/ : SUCCESS in 48s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/596/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/498/ : SUCCESS in 46s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/601/ : SUCCESS in 10s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/499/ : SUCCESS in 54s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/602/ : SUCCESS in 12s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/500/ : SUCCESS in 2m 15s

**Comment by gdgate** [comment]
Merge Failed.

This change or one of its cross-repo dependencies was unable to be automatically merged with the current state of its repository. Please rebase the change and upload a new patchset.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/604/ : SUCCESS in 10s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/503/ : SUCCESS in 51s

File: examples/src/components/Error.jsx [hunk 1]
```diff
@@ -23,8 +25,7 @@ export const Error = ({ error }) => (
             }}
         >
             <div className="gd-message-text">
-                <strong>{error.status}</strong>
-                { error.message ? <div>{error.message}</div> : null }
+                <strong>{JSON.stringify(error)}</strong>
```
Comment (zbycz) [inline examples/src/components/Error.jsx]: Sometimes the error is `{status: string}`, sometimes `{status {}}` - see `/basic-components`

It should be fixed in react-components, or at least JIRA task created.

Comment (zbycz) [reply examples/src/components/Error.jsx]: RAIL-670

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/607/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/505/ : SUCCESS in 46s

File: .gitignore [hunk 2]
```diff
@@ -38,6 +38,3 @@ coverage
 junit.xml
 
 test-storybook/
-
-.testcaferc.json
-/examples/_screenshots
```
Comment (zbycz) [inline .gitignore]: moved to /examples/.gitignore

File: examples/.eslintrc [hunk 3]
```diff
@@ -1,22 +1,30 @@
...
+        "jsx-a11y/click-events-have-key-events": 0,
+        "jsx-a11y/mouse-events-have-key-events": 0,
+        "object-curly-newline": 0,
+        "prefer-destructuring": 0,
+        "prefer-promise-reject-errors": 0,
+        "react/default-props-match-prop-types": 0,
+        "react/jsx-curly-brace-presence": 0,
+        "react/no-typos": 0,
 
+        "newline-per-chained-call": 0,
```
Comment (zbycz) [inline examples/.eslintrc]: This configuration didnt work... :(
`"newline-per-chained-call": ["error", { "ignoreChainWithDepth": 2 }]`

File: examples/src/components/ExecuteExample.jsx [hunk 4]
```diff
@@ -107,7 +107,7 @@ export class ExecuteExample extends Component {
         return (<div>
             {status}
             <p><button onClick={this.retry} className="button button-secondary s-retry-button">Retry</button> (fails every second attempt)</p>
-            {/* We need to render the Execute component even during loading otherwise the ongoing request is cancelled */}
+            {/* We need to render the Execute component even in loading otherwise the ongoing request is cancelled */}
```
Comment (zbycz) [inline examples/src/components/ExecuteExample.jsx]: max-line-length :-)

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/614/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/508/ : SUCCESS in 47s

File: examples/test/login_test.js [hunk 5]
```diff
@@ -11,5 +11,6 @@ test('should show login overlay and log in successfully', async (t) => {
         .typeText('input[name=email]', config.username)
         .typeText('input[name=password]', config.password)
         .click('button.s-login-submit')
-        .expect(loginOverlay.exists).notOk();
+        .expect(loginOverlay.exists)
+        .notOk();
```
Comment (zbycz) [inline examples/test/login_test.js]: Mistake - please take back when squashing..

File: examples/src/components/AttributeElementsExample.jsx [hunk 6]
```diff
@@ -19,74 +19,66 @@ export class AttributeFilterItem extends Component {
...
-        return (<div style={{ minHeight: 500 }}>
-            <AttributeElements
-                {...attributeElementsProps}
-            />
-        </div>);
+    render() {
+        return (
+            <div style={{ minHeight: 500 }}>
+                <AttributeElements identifier={employeeNameIdentifier} projectId={projectId} options={{ limit: 20 }}>
+                    {this.getInnerComponent}
```
Comment (vaclavbohac) [inline examples/src/components/AttributeElementsExample.jsx]: I would expect this to be `this.getInnerComponent()`

File: examples/src/components/LoginOverlay.jsx [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+                        <h1 style={{ textAlign: 'center', padding: 10 }}>{title}</h1>
+                        <div className="gd-input" style={{ margin: '5px 0' }}>
+                            <label htmlFor="email">e-mail</label>
+                            <input className="gd-input-field s-login-input-username" type="email" name="email" value={this.state.username} onChange={this.onUsernameChange} />
+                        </div>
+                        <div className="gd-input" style={{ margin: '5px 0' }}>
+                            <label htmlFor="password">password</label>
+                            <input className="gd-input-field s-login-input-password" type="password" name="password" value={this.state.password} onChange={this.onPasswordChange} />
+                        </div>
+                        { this.state.error && this.renderError() }
```
Comment (vaclavbohac) [inline examples/src/components/LoginOverlay.jsx]: Shouldn't the spaces around `this.state.error && this.renderError()` be caught by eslint rules?

File: examples/test/AttributeFilter_test.js [hunk 8]
```diff
@@ -0,0 +1,39 @@
+import { Selector } from 'testcafe';
+import { config } from './utils/config';
+import { loginUsingGreyPages } from './utils/helpers';
+
+fixture('Attribute filter components')
+    .beforeEach(loginUsingGreyPages(`${config.hostname}/attribute-filter-components`))
+    .page(`${config.hostname}/attribute-filter-components`);
+
+test('Dropdown filter', async (t) => {
```
Comment (vaclavbohac) [inline examples/test/AttributeFilter_test.js]: Could you update the title of the test so it's saying clearly what is it testing?

File: examples/test/AttributeFilter_test.js [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+        .click(clearButton)
+        .expect(attributeFilterItems.nth(0).child('input[type=checkbox]').checked).notOk()
+        .click(selectAllButton)
+        .expect(attributeFilterItems.nth(0).child('input[type=checkbox]').checked).ok()
+        .click(applyButton)
+        .expect(dropdownButton.hasClass('is-dropdown-open')).notOk()
+        .expect(dropdownButton.hasClass('is-active')).notOk();
+});
+
+test('Custom filter', async (t) => {
```
Comment (vaclavbohac) [inline examples/test/AttributeFilter_test.js]: Could you update the title of the test so it's saying clearly what is it testing?

File: examples/webpack.config.js [hunk 10]
```diff
@@ -11,43 +11,23 @@ process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';
 module.exports = ({ gdc = 'https://secure.gooddata.com', link = false } = {}) => {
     const proxy = {
         '/gdc': {
-            changeOrigin: true,
-            cookieDomainRewrite: 'localhost',
             target: gdc,
             secure: false,
+            cookieDomainRewrite: '',
             onProxyReq: (proxyReq) => {
                 // Browers may send Origin headers even with same-origin
```
Comment (vaclavbohac) [inline examples/webpack.config.js]: Is this comment still valid?

File: examples/test/AttributeFilter_test.js [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    const attirbuteFilterItem = Selector('.s-attribute-filter-list-item');
```
Comment (vaclavbohac) [inline examples/test/AttributeFilter_test.js]: `attribute`

File: examples/src/components/Header.jsx [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+class Header extends React.Component {
+    render() {
+        const { match, navigation, title } = this.props;
+        const href = match.path;
+        const navItem = href !== undefined ? navigation.find(link => link.href === href) : {};
+        const pageTitle = title !== undefined ? title : navItem.title;
+
+        const navigationElements = navigation.map(link => (
+            <li key={link.href}>
+                <NavLink to={link.href} className={'gd-header-menu-item s-menu-dashboards'} activeClassName={'active'} exact>
```
Comment (vaclavbohac) [inline examples/src/components/Header.jsx]: `className` could be just string here

File: examples/src/components/AttributeElementsExample.jsx (hunk 6 above)
Comment (tomasreichmann) [reply examples/src/components/AttributeElementsExample.jsx]: AttributeElements uses a children function, not an element, so this.getInnerComponent is correct, imho.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/637/ : SUCCESS in 7s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/521/ : SUCCESS in 45s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/638/ : SUCCESS in 10s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/522/ : SUCCESS in 47s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/639/ : SUCCESS in 13s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/523/ : SUCCESS in 1m 56s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/643/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/525/ : SUCCESS in 52s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/644/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/526/ : SUCCESS in 46s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/645/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/527/ : SUCCESS in 47s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-merge-compliance-check/149/ : SUCCESS in 27s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/528/ : SUCCESS in 46s
- https://ci.intgdc.com/job/gooddata-react-components-storybook-zuul-docker/157/ : SUCCESS in 2m 15s

**Pull request merged**